### PR TITLE
[DPO] Adding weighted preference optimization (WPO)

### PR DESCRIPTION
**Pull request by gaetanlop**
# What does this PR do?

Adding [WPO](https://huggingface.co/papers/2406.11827) to the `DPOTrainer`. The paper is in the list of accepted papers of [EMNLP 2024](https://2024.emnlp.org/program/accepted_main_conference/) and introduces an elegant method to similuate on-policy data while training on off policy preference pairs. It works by prioritizing the most probable samples under the current policy during optimization.

Implementation wise, it is independent from the loss function being used as it only provides a weighting for each sample pair during loss computation. My implementation is based on the authors implemation in https://github.com/wzhouad/WPO/blob/main/scripts/run_wpo.py

## Before submitting
- [X] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Did you write any new necessary tests?

## Who can review?

@kashif

### CODE REVIEW

**Review by kashif — APPROVED** [review]

**Comment by kashif** [comment]
@gaetanlop  tests for visual-llm models are failing i believe due to an extra argument?

**Comment by gaetanlop** [comment]
@kashif my bad, should be fixed now

**Comment by gaetanlop** [comment]
Hey @kashif, I don’t think the failing test is related to the PR. Am I right?

**Comment by qgallouedec** [comment]
> @gaetanlop tests for visual-llm models are failing i believe due to an extra argument?

No, see https://github.com/huggingface/trl/pull/2147#issuecomment-2385195093

**Comment by qgallouedec** [comment]
Please don't merge before #2131, it introduces some modification in the documentation, that I'd like this PR to take into account. I'll review this one in details when #2131 is merged

**Comment by gaetanlop** [comment]
Looks like #2131 has been merged. I will check if there are any modifications required.

**Comment by gaetanlop** [comment]
Thanks for the great doc @kashif, after checking #2131, it looks like it doesn't require any updates on this PR
